Mathlib/LinearAlgebra/LinearDisjoint.lean

```diff
@@ -0,0 +1,729 @@
+assuming `S` is commutative.
+
+## Tags
+
+linearly disjoint, linearly independent, tensor product
+
+-/
+
+open scoped Classical TensorProduct
+
+noncomputable section
+
+universe u v w
+
+namespace Submodule
+
+variable {R : Type u} {S : Type v}
+
+section Semiring
+
+variable [CommSemiring R] [Semiring S] [Algebra R S]
+
+variable (M N : Submodule R S)
+
+section mulMap
+
+-- can't use `LinearMap.mul' R S ∘ₗ TensorProduct.mapIncl M N` since it is not defeq to
+-- `(Algebra.TensorProduct.productMap A.val B.val).toLinearMap`
+
+/-- If `M` and `N` are submodules in an algebra `S` over `R`, there is the natural map
+`M ⊗[R] N →ₗ[R] S` induced by multiplication in `S`. -/
+def mulMap := TensorProduct.lift <| ((LinearMap.domRestrict' N).comp <| .mul R S).domRestrict M
+
+@[simp]
+theorem mulMap_tmul (m : M) (n : N) : mulMap M N (m ⊗ₜ[R] n) = m.1 * n.1 := rfl
+
+theorem mulMap_op :
+    mulMap (equivOpposite.symm (MulOpposite.op M)) (equivOpposite.symm (MulOpposite.op N)) =
+    (MulOpposite.opLinearEquiv R).toLinearMap ∘ₗ mulMap N M ∘ₗ
+    (TensorProduct.congr
+      (LinearEquiv.ofSubmodule' (MulOpposite.opLinearEquiv R).symm M)
+      (LinearEquiv.ofSubmodule' (MulOpposite.opLinearEquiv R).symm N) ≪≫ₗ
+    TensorProduct.comm R M N).toLinearMap :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comm_of_commute (hc : ∀ (m : M) (n : N), Commute m.1 n.1) :
+    mulMap N M = mulMap M N ∘ₗ TensorProduct.comm R N M := by
+  refine TensorProduct.ext' fun n m ↦ ?_
+  simp_rw [LinearMap.comp_apply, LinearEquiv.coe_coe, TensorProduct.comm_tmul, mulMap_tmul]
+  exact (hc m n).symm
+
+theorem mulMap_comp_rTensor (M' : Submodule R S) (hM : M' ≤ M) :
+    mulMap M N ∘ₗ (inclusion hM).rTensor N = mulMap M' N :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comp_lTensor (N' : Submodule R S) (hN : N' ≤ N) :
+    mulMap M N ∘ₗ (inclusion hN).lTensor M = mulMap M N' :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comp_map_inclusion (M' N' : Submodule R S) (hM : M' ≤ M) (hN : N' ≤ N) :
+    mulMap M N ∘ₗ TensorProduct.map (inclusion hM) (inclusion hN) = mulMap M' N' :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_eq_mul'_comp_mapIncl : mulMap M N = .mul' R S ∘ₗ TensorProduct.mapIncl M N :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_range : LinearMap.range (mulMap M N) = M * N := by
+  refine le_antisymm ?_ (mul_le.2 fun m hm n hn ↦ ⟨⟨m, hm⟩ ⊗ₜ[R] ⟨n, hn⟩, rfl⟩)
+  rintro _ ⟨x, rfl⟩
+  induction x using TensorProduct.induction_on with
+  | zero => rw [_root_.map_zero]; exact zero_mem _
+  | tmul a b => exact mul_mem_mul a.2 b.2
+  | add a b ha hb => rw [_root_.map_add]; exact add_mem ha hb
+
+/-- If `M` and `N` are submodules in an algebra `S` over `R`, there is the natural map
+`M ⊗[R] N →ₗ[R] M * N` induced by multiplication in `S`,
+which is surjective (`Submodule.mulMap'_surjective`). -/
+def mulMap' := (LinearEquiv.ofEq _ _ (mulMap_range M N)).toLinearMap ∘ₗ (mulMap M N).rangeRestrict
```

Comment:
```suggestion
def mulMap' : M ⊗[R] N →ₗ[R] ↥(M * N) :=
  (LinearEquiv.ofEq _ _ (mulMap_range M N)).toLinearMap ∘ₗ (mulMap M N).rangeRestrict
```
Does this work?

Reply:
In general please add a type ascription if possible and add a note explaining the reason otherwise.